src/dotty/tools/dotc/transform/TypeSpecializer.scala

```diff
@@ -0,0 +1,284 @@
+      defn.UnitType
+    )
+
+  private def defn(implicit ctx:Context) = ctx.definitions
+
+  private val specializationRequests: mutable.HashMap[Symbols.Symbol, List[Type]] = mutable.HashMap.empty
+  private val genericToInstantiation: mutable.HashMap[Symbols.Symbol, Type] = mutable.HashMap.empty
+  /**
+   *  A map that links symbols to their specialized variants.
+   *  Each symbol maps to another map, from the list of specialization types to the specialized symbol.
+   */
+  private val newSymbolMap: mutable.HashMap[Symbol, mutable.HashMap[List[Type], Symbols.Symbol]] = mutable.HashMap.empty
+
+  def allowedToSpecialize(sym: Symbol, numOfTypes: Int)(implicit ctx: Context): Boolean = {
+    (maxTparamsToSpecialize == 0 || numOfTypes <= maxTparamsToSpecialize) &&
+      numOfTypes > 0 &&
+      sym.name != nme.asInstanceOf_ &&
+      sym.name != nme.isInstanceOf_ &&
+      !(sym is Flags.JavaDefined) &&
+      !sym.isConstructor
+  }
+
+  def getSpecTypes(method: Symbol, poly: PolyType)(implicit ctx: Context): List[Type] = {
+    val requested = specializationRequests.getOrElse(method, List.empty)
+    if (requested.nonEmpty) requested
+    else {
+      if (ctx.settings.Yspecialize.value == "all") primitiveTypes.filter(tpe => poly.paramBounds.forall(_.contains(tpe)))
+      else Nil
+    }
+  }
+
+  def requestedSpecialization(decl: Symbol)(implicit ctx: Context): Boolean =
+    specializationRequests.contains(decl) ||
+      (ctx.settings.Yspecialize.value != "" && decl.name.contains(ctx.settings.Yspecialize.value)) ||
+      ctx.settings.Yspecialize.value == "all"
+
+  def registerSpecializationRequest(method: Symbols.Symbol)(arguments: List[Type])(implicit ctx: Context) = {
+    if (ctx.phaseId > this.treeTransformPhase.id)
+      assert(ctx.phaseId <= this.treeTransformPhase.id)
+    val prev = specializationRequests.getOrElse(method, List.empty)
+    specializationRequests.put(method, arguments ::: prev)
+  }
+
+  override def transformInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type = {
+    def generateSpecializations(remainingTParams: List[Name], specTypes: List[Type])
+                               (instantiations: List[Type], names: List[String], poly: PolyType, decl: Symbol)
+                               (implicit ctx: Context): List[Symbol] = {
+      if (remainingTParams.nonEmpty) {
+        specTypes.map(tpe => {
+          generateSpecializations(remainingTParams.tail, specTypes)(tpe :: instantiations, specialisedTypeToSuffix(ctx)(tpe) :: names, poly, decl)
+        }).flatten
+      }
+      else {
+        generateSpecializedSymbols(instantiations.reverse, names.reverse, poly, decl)
+      }
+    }
+    def generateSpecializedSymbols(instantiations: List[Type], names: List[String], poly: PolyType, decl: Symbol)
+                                  (implicit ctx: Context): List[Symbol] = {
+      val newSym =
+        ctx.newSymbol(decl.owner, (decl.name + "$mc" + names.mkString + "$sp").toTermName,
```

Comment:
If I understand the name mangling of specialization correctly, this should be `(decl.name + "$m" + names.mkString + "c$sp")`.
 `m` means method and `c` means class, for example if there's a specialized Int method in a specialized Int class, the signature will contain `$mIcI$sp`.